### PR TITLE
Add proper clasees to highlights

### PR DESCRIPTION
**Pull request by PiotrKozlowski**
for: https://app.zenhub.com/workspaces/openstax-unified-5b71aabe3815ff014b102258/issues/openstax/unified/864

- add "first" class to the first highlight's span
- add "last" class to the last highlight's span (if there is only one span then it will have both of these classes)
- add "text" or "block" classes to all of the highlight's spans ("block" class will be added for span containing img, video or iframe and "text" class will be added to the rest)
- allow highlighting video and iframes

### CODE REVIEW

File: src/injectHighlightWrappers.ts [hunk 1]
```diff
@@ -18,6 +18,12 @@ const IGNORE_TAGS = [
   'SCRIPT', 'STYLE', 'SELECT', 'OPTION', 'BUTTON', 'OBJECT', 'APPLET',
   'VIDEO', 'AUDIO', 'CANVAS', 'EMBED', 'PARAM', 'METER', 'PROGRESS',
 ];
+/**
+ * Highlights can be created around these block and text elements.
+ */
+const BLOCK_ELEMENTS = ['img', 'video', 'iframe'];
+const TEXT_ELEMENTS = ['.MathJax'];
+const ALLOWED_ELEMENT = BLOCK_ELEMENTS.concat(TEXT_ELEMENTS);
```
Comment (janKozie1) [inline src/injectHighlightWrappers.ts]: not sure if `video` should be both in ignored and allowed tags.

File: src/injectHighlightWrappers.ts [hunk 2]
```diff
@@ -116,7 +153,7 @@ function highlightRange(range: Range, wrapper: HTMLElement) {
   do {
     if (!node) { done = true; }
 
-    if (dom(node).matches('.MathJax,img')) {
+    if (dom(node).matches(ALLOWED_ELEMENT.join(','))) {
```
Comment (janKozie1) [inline src/injectHighlightWrappers.ts]: you could move the `.join` out of the loop

File: src/injectHighlightWrappers.ts (hunk 1 above)
Comment (PiotrKozlowski) [reply src/injectHighlightWrappers.ts]: Nicely spotted :D

File: src/injectHighlightWrappers.ts [hunk 3]
```diff
@@ -16,8 +16,14 @@ const NODE_TYPE = {
  */
 const IGNORE_TAGS = [
   'SCRIPT', 'STYLE', 'SELECT', 'OPTION', 'BUTTON', 'OBJECT', 'APPLET',
-  'VIDEO', 'AUDIO', 'CANVAS', 'EMBED', 'PARAM', 'METER', 'PROGRESS',
+  'AUDIO', 'CANVAS', 'EMBED', 'PARAM', 'METER', 'PROGRESS',
 ];
+/**
+ * Highlights can be created around these block and text elements.
+ */
+const BLOCK_ELEMENTS = ['img', 'video', 'iframe'];
```
Comment (TomWoodward) [inline src/injectHighlightWrappers.ts]: as far as i can tell the videos are also iframes. eg: https://rex-web.herokuapp.com/books/college-physics/pages/2-introduction-to-one-dimensional-kinematics so i'm not sure you have to handle them separately

Comment (TomWoodward) [reply src/injectHighlightWrappers.ts]: i ran a search and found no books with actual `video` elements

File: src/injectHighlightWrappers.ts [hunk 4]
```diff
@@ -80,9 +88,38 @@ function normalizeHighlights(highlights: Node[]) {
     }
   });
 
+  for (const [index, node] of (normalizedHighlights as HTMLElement[]).entries()) {
```
Comment (TomWoodward) [inline src/injectHighlightWrappers.ts]: you could put this type on the declaration on line 65

File: src/injectHighlightWrappers.ts [hunk 5]
```diff
@@ -80,9 +88,38 @@ function normalizeHighlights(highlights: Node[]) {
...
+ * Check if there are block elements inside node.
+ * Block elements are: img, video and iframe.
+ * @param {HTMLElement} node
+ * @returns {boolean}
+ */
+function hasBlockContent(node: HTMLElement) {
+  for (const block of BLOCK_ELEMENTS) {
+    if (node.querySelector(block)) { return true; }
+  }
+  return false;
```
Comment (TomWoodward) [inline src/injectHighlightWrappers.ts]: this whole thing could be `return !!node.querySelector(BLOCK_ELEMENTS.join(','));`

**Pull request merged**